rllib/examples/rl_modules/classes/inverse_dynamics_model_rlm.py

```diff
@@ -0,0 +1,252 @@
+    3) Predicting the next embedded obs, given an obs and action
+    ("forward" network).
+
+    The less the agent is able to predict the actually observed next feature
+    vector, given obs and action (through the forwards network), the larger the
+    "intrinsic reward", which will be added to the extrinsic reward.
+    Therefore, if a state transition was unexpected, the agent becomes
+    "curious" and will further explore this transition leading to better
+    exploration in sparse rewards environments.
+
+    .. testcode::
+
+            import numpy as np
+            import gymnasium as gym
+            import torch
+
+            from ray.rllib.core import Columns
+            from ray.rllib.core.rl_module.rl_module import RLModuleConfig
+            from ray.rllib.examples.rl_modules.classes.inverse_dynamics_model_rlm import (  # noqa
+                InverseDynamicsModel
+            )
+
+            B = 10  # batch size
+            O = 4  # obs (1D) dim
+            A = 2  # num actions
+            f = 25  # feature dim
+
+            # Construct the RLModule.
+            rl_module_config = RLModuleConfig(
+                observation_space=gym.spaces.Box(-1.0, 1.0, (O,), np.float32),
+                action_space=gym.spaces.Discrete(A),
+            )
+            icm_net = InverseDynamicsModel(rl_module_config)
+
+            # Create some dummy input.
+            obs = torch.from_numpy(
+                np.random.random_sample(size=(B, O)).astype(np.float32)
+            )
+            next_obs = torch.from_numpy(
+                np.random.random_sample(size=(B, O)).astype(np.float32)
+            )
+            actions = torch.from_numpy(
+                np.random.random_integers(0, A - 1, size=(B,))
+            )
+            input_dict = {
+                Columns.OBS: obs,
+                Columns.NEXT_OBS: next_obs,
+                Columns.ACTIONS: actions,
+            }
+
+            # Call `forward_train()` to get phi (feature vector from obs), next-phi
+            # (feature vector from next obs), and the intrinsic rewards (individual, per
+            # batch-item forward loss values).
+            print(icm_net.forward_train(input_dict))
+
+            # Print out the number of parameters.
+            num_all_params = sum(int(np.prod(p.size())) for p in icm_net.parameters())
+            print(f"num params = {num_all_params}")
+    """
+
+    @override(TorchRLModule)
+    def setup(self):
+        # Get the IDM achitecture settings from the our RLModuleConfig's (self.config)
+        # `model_config_dict` property:
+        cfg = self.config.model_config_dict
+
+        feature_dim = cfg.get("feature_dim", 288)
+
+        # Build the feature model (encoder of observations to feature space).
+        layers = []
+        dense_layers = cfg.get("feature_net_hiddens", (256, 256))
+        # `in_size` is the observation space (assume a simple Box(1D)).
+        in_size = self.config.observation_space.shape[0]
```

Comment:
This works only for 1d spaces, does it?

Reply:
Currently, yes. If users need e.g. image-obs support, they can now easily write their own ICMs. I didn't want to make the example ICM too complicated.